칸반 보드 아이템 이동 구현

ingong opened the pull request:
## 😀 제목

칸반 보드 아이템 이동 구현 

## ⛅️ 내용

> 이 PR의 작업 요약

- 정수 자리로 order 를 갖도록 수정완료
- id 를 DB 에서 받아온 이후에 할당하도록 수정 완료
- kanbanAddBtn 에 switch 문 로직 제거
- 칸반 Item 내부에 Input 컴포넌트 분리 (책임 분리)
- 칸반 아이템에 해당 아이템 가져갖을 때 파란색 줄이 나타나게 구현 완료(노션처럼)
- 아이템을 드랍했을 때 해당 위치를찾아가도록 구현완료
- dispatch 를 통한 상태 변경 이후에 story API 를 통해서 order DB 최신화

## 🎸특이사항

> 리뷰시 참고할만한 내용, 주의깊게 봐줬으면 하는 내용

- 아까 회고할 때 말씀해주셨던 post 를 하고 해당 item 을 반환하는 방법을 아직 못찾아서 일단 get 으로 가져오게 구현했습니다.
- 현재는 동일 컬럼내에서 이동이 가능합니다. 

- 마무리될지는 모르겠지만, 구현을 좀 더 해보고 내일 아침에 결과를 말씀드릴게요 👍 
- 크흐....드래그 앤 드랍 끝냈습니다. UX 향상을 위해 필요한 Event Handling 이 하나 정도 더 있을 것 같네요. 쪼갠다고 쪼갰는데 드래그 앤 드랍은 경우의 수가 크게 4가지가 나와서 내일 함수로 또 분리해봐야할 것 같아요. 테스트 코드는 리펙토링 이후에 작성해볼게요....! 여러분들의 0, 1, 2 로 id 를 작성해보라고 하셨던게 아주 저의 작업을 편하게 해주었어요. 감사합니다 :) 영상도 짧게 첨부해봅니다ㅎ

https://user-images.githubusercontent.com/73208317/142909133-e6dcaffc-9f0c-4005-be6f-82cee462a330.mov

ingong committed:
fix: 스토리 Order 프로퍼티 정수 자리 8, 소수 12 자리로 고정 [#87]

- order 의 자리수 수정
- 새로 story Item 이 생성될 때마다 1 씩 증가하는 값을 가지게함

ingong committed:
refactor: StoryItem Order 를 순차적으로 가지도록 수정 [#87]

ingong committed:
chore: storyType 에 order 를 undefined 가능하도록 수정 [#87]

ingong committed:
fix: story 의 id 를 DB 에서 자동 생성된 값을 받아오도록 Controller 함수 수정 [#87]

ingong committed:
feat: Story 를 id 로 가져오는 Controller 함수 작성 [#87]

ingong committed:
feat: Client 에서 id 로 story 를 조회하도록 함수 작성 [#87]

- getStoryByid 함수 작성
- server 에서 router id -> :id 로 변경함...!

ingong committed:
fix: storyType 수정, kanBanAddBtn 에서 Story 추가하는 로직 구현 [#87]

- storyType 의 id 타입을 undefined 가 가능하게 지정 (DB 에서 생성시 받아옴)
- addBtn 을 누르면 Story 가 생성되고, 이를 id 로 조회해서 dispatch 함

ingong committed:
refactor: story update 함수 id 전송을 body 에서 params 로 수정 [#87]

ingong committed:
chore: story order 기준으로 정렬하도록 타입 변환 [#87]

- object is possibly undefined 를 해결하기 위한 커밋
- storiesController TODO 작성

ingong committed:
feat: 칸반 아이템에 isDragEnter 상태를 추가함 [#87]

- dragEnter 의 상태에 따라서 밑에 부분에 파란색 줄을 보이기 위함

ingong committed:
layout: kanbanItem 이 dragEnter 상태일 때 파란색 줄 스타일링 [#87]

ingong committed:
feat: Drag 가 된 아이템에 대해서는 파란선이 사라지도록 작성 [#87]

ingong committed:
refactor: KanBanType 정리 및 기존 파일에서 사용하는 export 제거 [#87]

- 불필요한 export 를 제거하고, 하나의 파일에서 타입을 관리하기위함

ingong committed:
refactor: KanBanItem 과 KanBanItemInput 분리 [#87]

- KanBanItem 은 드래깅 하는 이벤트 관리할 예정이라 무거워질 것 같아서 분리

ingong committed:
layout: Kanban 과 KanbanItem 스타일 분리 [#87]

- 이전 커밋과 동일한 이유로 분리

ingong committed:
layout: 칸반 이동 맨 상단 부분 Enter 시에 파란색 줄 표시 [#87]

ingong committed:
feat: 같은 컬럼 내에서 이동 구현 완료 [#87]

haesoo-y commented on client/src/components/KanbanAddBtn/index.tsx:
```diff
@@ -1,49 +1,29 @@
-const extractLastID = (array: Array<StoryType>): number => {
-  if (!array?.length) return 0;
-  return Math.max(...array.map((v) => v.id));
+//TODO projectID, epicId 주입 예정
```
TODO tree 쓰시나요? 👍

haesoo-y commented on client/src/components/KanbanColumn/style.ts:
```diff
@@ -14,20 +18,31 @@ const Styled = {
+    padding-top: 12px;
+    padding-bottom: 8px;
+    font: ${({ theme }) => theme.font.bold_regular};
+    border-bottom: 4px solid transparent;
+    border-bottom: ${({ isTopEnter, theme }) => isTopEnter && `4px solid${theme.color.blue200}`};
```
border-bottom이 두 줄이네요...

haesoo-y commented on client/src/layers/Kanban/index.tsx:
```diff
@@ -18,13 +13,13 @@ const Kanban = () => {
   const finishList = storyList
     .filter((item) => item.status === 'DONE')
-    .sort((a, b) => a.order - b.order);
+    .sort((a, b) => Number(a.order) - Number(b.order));
```
```suggestion
    .sort(storySortByOrder(a,b));
```
반복되는 정렬함수를 빼보는 건 어떨까요?? `utils/sort.ts` 가 준비되어있습니다.😉

rudy3091 replied:
저도 찬성입니다! 다만.. sortStoryByOrder 로 해주시면 안될까요? 함수는 동사로 시작하는게 좋으니깐..

ingong committed:
feat: 동일 칼럼에서 최상단, 최하단으로 Item 이동 기능 구현 완료 [#87]

- 동일 컬럼 내에서 아이템 이동 기능 구현 완료

haesoo-y commented on server/src/Stories/Stories.controller.ts:
```diff
@@ -33,26 +33,40 @@ export const getAllStoriesByProject = async (req: Request, res: Response) => {
+    if (!result) {
+      throw new Error(`해당 ${id} 를 조회할 수 없습니다`);
+    }
+    res.status(200).json(result);
```
express res.json의 디폴트 값이 status 코드 200인 것으로 알고 있습니다.
```suggestion
    res.json(result);
```

ingong committed:
refactor: dragEnter, dragStart 함수 분리 [#87]

- 칸반 Column 내에서 정의되는 함수가 많고, 이는 KanBanItem 에서 사용하기 때문

ingong committed:
refactor: 칸반 Item drag 함수 분리 및 타입 재선언 [#87]

ingong committed:
feat: 동일한 카테고리인지 확인하는 함수 작성 [#87]

- Column 기준으로 동일한지에 따라서 달라지기 때문에

rudy3091 commented on client/src/components/KanbanColumn/index.tsx:
```diff
@@ -1,59 +1,85 @@
-    dragOverRef.current = null;
+  const handleDragDrop = (category?: StatusType) => {
+    // 동일한 칼럼 내의 제일 상단에 Item 을 놓을 때
+    if (isTopEnter) {
```
여기에 얼리 리턴을 쓰시면 depth 를 하나 더 줄일 수 있을것같습니다!

rudy3091 commented on client/src/components/KanbanColumn/index.tsx:
```diff
@@ -1,59 +1,85 @@
+      <Styled.KanBanColumnTitle
+        onDragOver={(e) => e.preventDefault()}
+        onDragEnter={() => setTopEnter((isTopEnter) => !isTopEnter)}
+        onDragLeave={() => setTopEnter(false)}
```
오호 dragLeave 이벤트도 처리를 해줘야 되는군요...

사실 로직을 자세히 보진 못했습니다.. 제 한계인가봐요...ㅠㅠ 내일 설명 부탁드릴게요!

rudy3091 commented on client/src/components/KanbanItem/index.tsx:
```diff
@@ -1,50 +1,48 @@
+        handleDragStart(e, story.order as number, story.status, draggingRef, draggingCategory);
+        setDragEnter(false);
+      }}
+      onDragEnter={(e) => {
+        handleDragEnter(e, story.order as number, story.status, dragOverRef, dragOverCategory);
+        setDragEnter((isDragEnter) => !isDragEnter);
+      }}
+      onDrop={(e) => {
+        handleDragDrop(story.status);
+        setDragEnter(false);
+      }}
```
dragEnter 이랑 drop 이벤트 핸들러를 item 에 붙여주셨군요! 뭔가 제가 구현하고 나니깐 서로 어떤점이 다른지 눈에 들어오네요. 내일 설명이 더더욱 기대됩니다 🙂

rudy3091 commented on client/src/components/KanbanItem/style.ts:
```diff
@@ -1,27 +1,22 @@
 import styled from 'styled-components';
 import cancelicon from '@public/icons/cancel-icon.svg';
 
+interface KanbanItemPropType {
+  isDragEnter: boolean;
+}
```
client/src/components/KanbanColumn/style.ts 에도 쓰였던것 같은데 분리해서 함께 써도 되지 않을까요?? 그게 아니라면 inline 타이핑을 하는게 가독성엔 더 도움이 될 것 같다는 개인적인 의견입니다!

rudy3091 commented on client/src/components/KanbanItemInput/style.ts:
```diff
@@ -0,0 +1,16 @@
+  width: 90%;
+  height: 30px;
+
+  font: ${({ theme }) => theme.font.bold_small};
+  font-size: 14px;
```
font theme 안에 size 가 설정돼있는데 의도하신건가요??

ingong committed:
refactor: KanBan 에서 KanBan Column Props StoryList 제거 [#87]

- Context API 를 통해서 받아오도록함
- 나중에 filter 를 통해서 재사용해야하기 때문

ingong committed:
feat: 다른 컬럼으로 스토리 아이템 이동 기능 구현 완료 [#87]

- 아직 다른 컬럼의 최상단 부분으로 이동은 구현 중
- 생각보다 자연스러운 상태 변경이 된다 뿌듯...

rudy3091 commented on client/src/layers/Kanban/index.tsx:
```diff
@@ -5,41 +5,52 @@ import KanbanModal from '@/components/KanbanModal';
-            categoryRef={cateogryRef}
             category={'TODO'}
             storyList={todoList}
+            draggingRef={draggingRef}
+            dragOverRef={dragOverRef}
+            draggingCategory={draggingCategory}
+            dragOverCategory={dragOverCateogry}
+          />
+          <KanbanColumn
+            category={'IN_PROGRESS'}
+            storyList={onGoingList}
+            draggingRef={draggingRef}
+            dragOverRef={dragOverRef}
+            draggingCategory={draggingCategory}
+            dragOverCategory={dragOverCateogry}
+          />
+          <KanbanColumn
+            category={'DONE'}
+            storyList={finishList}
+            draggingRef={draggingRef}
+            dragOverRef={dragOverRef}
+            draggingCategory={draggingCategory}
+            dragOverCategory={dragOverCateogry}
```
dragging- 가 계속 반복되는데 이렇게 하면 어떨까요? 너무 억지인가요??

```suggestion
          [
            { category: 'TODO', storyList: todoList },
            { category: 'IN_PROGRESS', storyList: onGoingList },
            { category: 'DONE', storyList: finishList },
          ].map(({ category, storyList }) => (
            <KanbanColumn
              category={category}
              storyList={storyList}
              draggingRef={draggingRef}
              dragOverRef={dragOverRef}
              draggingCategory={draggingCategory}
              dragOverCategory={dragOverCateogry}
            />
          ));
```

rudy3091 commented on client/src/lib/api/story.ts:
```diff
@@ -17,32 +17,25 @@ export const getAllStories = async (projectId: number | string) => {
       epicId,
     });
-    return result.data;
+    return result.data.id;
```
해수님 PR에도 달았던 건데, 뭔가 새로운 항목을 생성하거나 수정하는 경우에는 toast.success로 피드백을 주는건 어떨까요? 인송님 PR을 먼저 볼수도 있어서 여기도 남겼습니다ㅋㅋㅋ

rudy3091 commented on client/src/types/story.d.ts:
```diff
@@ -1,10 +1,33 @@
+import { dragRefObjectType } from './story.d';
```
오잉 이건 뭔가요??ㅋㅋㅋㅋ 재귀 import 처럼 보이네요ㅋㅋㅋㅋㅋ

ingong committed:
feat: 다른 칼럼 내 최상단으로 Item 이동 기능 구현 완료 [#87]

- KanbanColumn 의 길이가 150줄이다... 내일 리펙토링해야지
- 그래도 함수 내에 조건문에 따른 구현이 명확해서 어렵지는 않을 것 같다

rudy3091 commented on server/src/Stories/Stories.router.ts:
```diff
@@ -1,18 +1,21 @@
+  .get('/', getAllStoriesByProject)
+  .get('/:id', getStoryById)
+  .post('/', postStory)
+  .patch('/name/:id', updateStoryWithName)
+  .patch('/order/:id', updateStoryWithId)
```
controller에서는 updateStoryWithId 와 updateStoryWithName 모두 id를 기준으로 변경하고 계시고 후자는 order를 추가로 받는것 밖에 없는것 같은데 두 엔드포인트를 통합하고 updateStoryWithName을 써도 되지 않을까요??

그럴 필요는 없지만 story의 다른 필드들을 모두 받아서 update 해주면 그만큼 오버헤드가 생기는걸까요?? 데이터베이스에 이것저것 실험해보고싶은데 기능을 구현하기도 시간이 없어서 슬프네요.. 😥

rudy3091 reviewed (commented):
고생 많으셨습니다!! 드래그앤 드랍... 어렵더라구요... 인송님 로직을 완전히 이해한건 아니지만 정말 어려운 작업이라는걸 느꼈습니다ㅎㅎㅎ

ingong committed:
feat: KanBanBoard 위치 변경 API 함수 호출 [#87]

- 각 상황별 API 를 호출하도록 수정함

ingong committed:
chore: type 파일 오타 수정

ChanHoLee275 commented on client/src/components/KanbanColumn/index.tsx:
```diff
@@ -1,59 +1,148 @@
+  const storyList = useStoryState();
+  const filterList = storyList
+    .filter((item) => item.status === category)
+    .sort((a, b) => Number(a.order) - Number(b.order));
```
order에 Number를 한 번 더 감싼 이유가 있을까요?

ChanHoLee275 commented on client/src/components/KanbanColumn/index.tsx:
```diff
@@ -1,59 +1,148 @@
+    }
+
+    // 동일한 컬럼 내의 Item 사이 또는 최하단
+    if (!isTopEnter && isEqualCategory(draggingCategory, category)) {
```
위의 조건문을 검사한 다음에 한 번 더 검사를 해야 하는 경우인가요?
아닌 경우면 else if를 쓰는 것이 좋을 것 같아요...!
